### PR TITLE
Feature/admission webhook improvements

### PR DESCRIPTION
**Pull request by kesselborn**
This PR adds some conveniences around the admission webhook -- everytime I needed to implement a webhook, I was quite annoyed about the many moving parts that have to be coordinated.
This looks like a lot of code but there is quite a bit of documentation and test code included + the derive code is more or less trivial but annoying to implement for each project individually.
It's probably easiest to look at each commit indivually as they are split up between `krator` and `krator_derive`.
This is still blocked until #574  (or the bug it that one tries to fix) is fixed, but perhaps people interested can already have a look at it.
/cc @kflansburg

### CODE REVIEW

**Comment by kflansburg** [comment]
Hey @kesselborn thanks for putting this together. Something like this was definitely on my roadmap. 

This will probably take some time to review, but I do have some questions. The thing that originally prevented me from doing something like this in the moose example is that it requires the operator itself to have RBAC permissions to make these configuration changes, which I don't believe is best practice. Does this require those elevated permissions? What are your thoughts on this? I had planned to implement some sort of tooling like a Cargo plugin to deploy the operator and make this configuration.

**Comment by kesselborn** [comment]
yes: if you want to do it the way I added it to the moose example, you would need elevated permissions ... but as it is only an example, I think that's ok + you can see how you can use those functions.
In a current operator I am writing, you can just output the crd when calling it with an option.

That's why I added the Type `WebhookResources` as a wrapper type and implemented the `Display` crate. This is analagous to the `crd()` command from the `kube_derive` macro ... so you can either just output the manifests as yaml or write code that installs those resources. The code can than be run locally or within the cluster.

With this approach, you have total flexibility when you provide an operator:
- you could just auto-generate the files in a pre-commit hook and add the yaml-files to your repository
- you could create a binary that installs the files and can be run once before deploying the operator
- you could auto-generate a very specific RBAC-file which just allows to install exactly these resources
- you could document how users can create the yaml files

... the quick win is that working code and documentation don't diverge.

While I like cargo plugins, this would make Rust a dependency for people who want to use the operator and I think Rust is not (yet) very common for many people. As an Operator-Provider, I would like people to be able to use the operator without the need to use something other than the normal Kubernetes-Tooling

**Comment by kesselborn** [comment]
I am currently thinking, whether it would make sense to extract the derive macro into it's own little crate. It's not really related to krator but could be used independently as well.
The only remaining part would be the extraction of the certificate and the private key -- what do you think @kflansburg 

I'll think about it and probably adjust the pull request accordingly

File: crates/krator-derive/src/admission.rs
Comment (kflansburg) [inline crates/krator-derive/src/admission.rs:106]: There is a lot of stuff here that I think we may want to be able to configure via attributes.

File: crates/krator/examples/moose.rs
Comment (kflansburg) [inline crates/krator/examples/moose.rs:398]: I think this would ideally be run automatically by `krator` when a given feature flag is enabled.

**Review by kflansburg — CHANGES_REQUESTED** [review]
I think this looks good, albeit very complex. IMO the best practice for this would be to emit manifests / kustomize + Makefile in a similar fashion to `kubebuilder`, with auto-apply being an opt-in feature flag. I do think this gets us closer to that so I'm in favor of merging after some of my comments are addressed. 


@thomastaylor312 I think you should take a look at this since it touches a lot of stuff in `krator-derive`. I'm not sure how common it is to have so many dependencies in a derive crate but it seems like an antipattern.

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml:45]: ah, probably not ... at first I only needed it in test and forgot to remove it

File: crates/krator/Cargo.toml
Comment (kesselborn) [reply crates/krator/Cargo.toml:44]: oh, no: this was probably a toml plugin in vscode

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs:306]: yeah, I am aware or that ... I thought it would be ok in an example ... could fix that though

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs:371]: yup ... again: for the example I made some choices that I wouldn't do for a proper project and I thought about whether this is ok. But I think you are right: hard-coding default is probably not a good idea.

As to being more secure when mounting the secret: yes, you are right. Again, for the example here, we don't really know, whether the example is being executed locally or within the cluster -- reading the secret from the API works in both cases. Perhaps a comment explaining the consequence would be sufficient here?

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs:387]: oh, nice one ... thanks. I was looking for this but didn't find it.

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs:106]: attributes to the feature? good idea! Can you switch on/off dependencies with this? The only real gain would probably less dependencies if, for example, you are not interested in the automatic generation of the certificate & private key.

File: crates/krator-derive/Cargo.toml
Comment (thomastaylor312) [inline crates/krator-derive/Cargo.toml:52]: This creates a circular dependency when trying to push the crate. We had this in here before (I think)

File: crates/krator-derive/src/admission.rs
Comment (thomastaylor312) [inline crates/krator-derive/src/admission.rs:52]: Nit (spelling): Admission

File: crates/krator-derive/src/admission.rs
Comment (thomastaylor312) [inline crates/krator-derive/src/admission.rs:216]: Are `unwrap`s the correct thing here? Seeing as these are errors, we should probably do an `unwrap_or_default`

**Comment by kesselborn** [comment]
@kflansburg sorry for the late reploy and thanks for taking the time to review this rather big PR -- very much appreciated.

Before fixing the comments I would like to discuss again, if it makes sense to make this part of `krator_derive`. I implemented it there as I was using `krator` and was missing the functionality. When thinking about it again however, I think none of the additions are related to `krator` and could be used as well when writing an operator w/o using `krator` ... whereas pulling in `krator_derive` without using `krator` makes little sense.

I think extracting this as an individual crate is probably cleaner? What do you think?

Regarding pulling in a lot of additional dependencies via a proc macro: I was wondering this myself as well, and would be interested in @thomastaylor312 opinion here as well.

File: crates/krator-derive/src/admission.rs
Comment (thomastaylor312) [inline crates/krator-derive/src/admission.rs:209]: I think all of the error types should be `Box<dyn Error + Send + Sync>` for maximum capability. We do not want to force people deriving things to use `anyhow`. If we already are with krator, then this is fine for now

File: crates/krator-derive/src/lib.rs
Comment (thomastaylor312) [inline crates/krator-derive/src/lib.rs:9]: This should probably contain the list of dependencies the generated code uses, which I think are `kube`, `k8s_openapi`, and `anyhow`.

File: crates/krator-derive/Cargo.toml
Comment (thomastaylor312) [inline crates/krator-derive/Cargo.toml:27]: After reading through the new code, I don't think we need most of these new dependencies. The code in the macro is dependent on the project using `krator-derive` to import the correct dependencies

File: crates/krator-derive/src/lib.rs
Comment (thomastaylor312) [inline crates/krator-derive/src/lib.rs:46]: This is a great example! The one problem (like in a previous comment) that this code cannot depend on `krator` or we get a weird circular dependency issue. If we can get it so that isn't needed in `dev-dependencies`, this will work fine. If it does need a dependency on `krator`, we'll want to move this example there

File: crates/krator/examples/moose.rs
Comment (thomastaylor312) [reply crates/krator/examples/moose.rs:306]: Yeah, I think we can just pop an allow unused on there or conditionally include this parameter when the admission webhook feature isn't enabled

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
This is looking pretty good. Thank you for adding tests and cleaning up the derive stuff.! Insofar as I understand, I don't think we need to import any `dependencies` for the derive macros to work, but I could be wrong

**Comment by kflansburg** [comment]
@kesselborn Given that `krator-derive` should not depend on `krator`, does that address your concerns about splitting it out into a new crate? I think for now it makes most sense to me in `krator-derive`. I think this could still be used without `krator`.

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml:28]: [the default feature settings](https://github.com/deislabs/krustlet/blob/7c47872f33108b05d95c126c7820a2882dd7d68b/crates/krator-derive/Cargo.toml#L37) does this, doesn't it?

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml]: fixed in commit acc48b7

File: crates/krator-derive/src/lib.rs
Comment (kesselborn) [reply crates/krator-derive/src/lib.rs:10]: these details are available if you click on the link [AdmissionWebhook]

image

I can move it to crate level as well -- not sure what is more idiomatic

File: crates/krator-derive/tests/admission.rs
Comment (kesselborn) [reply crates/krator-derive/tests/admission.rs]: smuggled it into commit 9362f6f

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs:306]: I changed the code in commit 9362f6f and this change is not necessary anymore

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs]: made namespace configurable in commit 926d048

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs]: fixed in commit 5afaea5

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs]: I don't think I really agree here ... there are many different way of installing CRD and the other resources, especially due to different RBAC permissions. The function itself is kind of trivial when used like that and as it ships within the example, users can copy it if they like the approach.

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml]: good catch ... although I think it's fine as it is only in `dev-dependencies`.
I removed it as it is not needed anymore (it was part of a test before)

removed in commit c3391b9

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs]: fixed in commit 6e5fd8b

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs:208]: 🤯 I got to admit, that I don't understand this as I am kind of new to Rust ... if you give me a pointer, I am happy to adjust the code accordingly

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs]: fixed in commit 5d54163 ... I am catching errors early with `ensure!` to have more speaking error messages

**Comment by kesselborn** [comment]
@kflansburg regarding the splitting out into a separate crate: the transition macro creates a 

```rust
impl krator::TransitionTo
```

block ... doesn't that make it kind of bound to krator?
However: happy to stand corrected, no too familiar what is most common

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml:27]: yeah ... my hope was that the dependencies get pulled in as a transient dependency. 
Is there a good way to solve this? For example: there is a method creating a secret with a certificate and a private key using the `rcgen` crate. If this is not present, the user of `krator-derive` will get an error that `rcgen` is missing pointing to the derive macro ... is this something that users of macros usually have to solve themselves and is simply documented?

File: crates/krator-derive/src/lib.rs
Comment (kesselborn) [reply crates/krator-derive/src/lib.rs:9]: that makes sense, yes

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs:105]: I will have to look into how to parse those attributes

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml:28]: ok: just saw the windows error: 

```
   Compiling num_cpus v1.13.0
error: failed to run custom build command for `openssl-sys v0.9.60`

Caused by:
  process didn't exit successfully: `D:\a\krustlet\krustlet\target\debug\build\openssl-sys-8ab17b7cd200101e\build-script-main` (exit code: 101)
  --- stdout
```

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml]: I cleaned up the dependecies in commit dec3236

File: crates/krator-derive/src/admission.rs
Comment (thomastaylor312) [reply crates/krator-derive/src/admission.rs:208]: Hahah no problem. So when you are trying to implement errors in the most generic way possible (meaning that you aren't going to require the use of a specific library), you can change return type to `Result<k8s_openapi::api::admissionregistration::v1::MutatingWebhookConfiguration, Box<dyn Error + Send + Sync>>`. Boxing a type that implements a trait in rust is how dynamic dispatch works but it is also essentially the "generic error type" or lowest common denominator in Rust. We also require the `Send + Sync` trait bounds as all of these are being used in an async context and will likely need to be sent across `await`s.

Hopefully that makes more sense? Honestly, if you don't feel comfortable doing it, we can handle it in a follow up too.

File: crates/krator-derive/Cargo.toml
Comment (thomastaylor312) [reply crates/krator-derive/Cargo.toml]: Yeah it is generally documented and not pulled into a derive crate. Insofar as I understand things, derive macros are purely meant as transparent code generation

**Comment by thomastaylor312** [comment]
To clarify, it doesn't depend on krator in the Cargo dependency sense. I think my comment [here](https://github.com/deislabs/krustlet/pull/577#discussion_r617690699) explains that a bit more

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs:180]: this is implemented in commit 193b6ac

File: crates/krator-derive/src/lib.rs
Comment (kesselborn) [reply crates/krator-derive/src/lib.rs:9]: done in commit 193b6ac

File: crates/krator-derive/src/admission.rs
Comment (kesselborn) [reply crates/krator-derive/src/admission.rs]: @thomastaylor312 thanks for taking the time explaining this. That is kind of what I thought this is about. My remaining problem was on how to create this kind of error.
It took me a while to find out but finally a colleague poked me to `Err("foo".into())`  ... this should be mentioned in every Rust-error handling tutorial / texts / docs ;) 

I did this in commit e8a64e2

File: crates/krator-derive/src/lib.rs
Comment (kesselborn) [reply crates/krator-derive/src/lib.rs:10]: I am resolving this for now, feel free to unresolve if you disagree.
The only thing that we have to make sure is that the feature `admission-webhook` is used when creating the docs -- can this be configured somehow?

File: crates/krator-derive/src/lib.rs
Comment (kesselborn) [reply crates/krator-derive/src/lib.rs:77]: is it really a problem in `dev-dependencies` as well? I got problems when having circular dependencies in `dependencies` but no complains in `dev-dependencies` ... but that's perhaps just a cheap `works for me` ;)

File: crates/krator-derive/Cargo.toml
Comment (kesselborn) [reply crates/krator-derive/Cargo.toml]: these dependencies are removed now and the Windows build works

**Comment by kesselborn** [comment]
@kflansburg / @thomastaylor312 : can we move this forward somehow? I know, it's quite a lot of comments and therefore some work to review.

I resolved all comments apart from the ones, where I had a follow-up

**Comment by kflansburg** [comment]
> @kflansburg / @thomastaylor312 : can we move this forward somehow? I know, it's quite a lot of comments and therefore some work to review.
> 
> 
> 
> I resolved all comments apart from the ones, where I had a follow-up

Apologies, I've been very busy this last week. I will try to take a final look at this soon.

**Comment by bacongobbler** [comment]
Hey so I'm trying to run the examples, but I'm running into an issue with the admission webhook being unreachable:

```
><> k create -f manfred.yaml
Error from server (InternalError): error when creating "manfred.yaml": Internal error occurred: failed calling webhook "mooses.animals.com": Post "https://mooses-animals-com-admission-webhook.default.svc:443/?timeout=10s": dial tcp 10.96.19.121:443: connect: connection refused
```

I'm trying to follow the guide as directed in the logs:

```
May 05 09:36:15.807 moose: 
    
Running moose example. Try to install some of the manifests provided in examples/assets
```

This is with a KinD cluster, in case that matters. I'll try with AKS and see if my results differ.

**Comment by bacongobbler** [comment]
Oddly I receive a different result with AKS:

```
><> k create -f manfred.yaml
Error from server (InternalError): error when creating "manfred.yaml": Internal error occurred: failed calling webhook "mooses.animals.com": Post "https://mooses-animals-com-admission-webhook.default.svc:443/?timeout=10s": read unix @->/socket/konnectivity: read: connection reset by peer
```

**Comment by bacongobbler** [comment]
Things work just fine without the admission webhook feature enabled.

```
><> k create -f manfred.yaml
moose.animals.com/manfred created
><> k create -f melissa.yaml
moose.animals.com/melissa created
```

```
Running moose example. Try to install some of the manifests provided in examples/assets
    
    
    at crates/krator/examples/moose.rs:529

  May 05 10:36:29.473 krator::runtime: Got a watch restart. Resyncing queue...
    at crates/krator/src/runtime.rs:297

  May 05 10:36:29.473 krator::runtime: Finished resync of objects.
    at crates/krator/src/runtime.rs:300

  May 05 10:36:53.484 moose: Found new moose named manfred!
    at crates/krator/examples/moose.rs:127

  May 05 10:37:00.263 moose: Found new moose named melissa!
    at crates/krator/examples/moose.rs:127
```

**Comment by kesselborn** [comment]
hey @bacongobbler : do you run the moose example outside the cluster or do you deploy it into the cluster? If you run it outside the cluster, you need to do some "hacking", so that the example can process the mutation requests ... if you look further up in the log it should have respective instructions:

```
If you run this example outside of Kubernetes (i.e. with `cargo run`), you need to make the webhook available.
Try the script example/assets/use-external-endpoint.sh to redirect webhook traffic to this process. If this
operator runs within Kubernetes and you use the webhook resources provided by the admission-webhook macro, 
make sure your deployment has the following labels set: ...
```

I tested locally using kind as well ... if you test it with AKS, you would have to make sure, that the Kubernets-Cluster can talk to your process

**Comment by kesselborn** [comment]
so, just to give more context: these admission-webhooks are really just that: webhooks. So, in order to provide a webhook, the example app provides an https-endpoint where Kubernetes will post `moose` manifests whenever you create a new `moose`. This endpoint needs to be available through the service `mooses-animals-com-admission-webhook`. When you run the executable outside the cluster, you need to make the endpoint available to this service ... this is what the script ` script example/assets/use-external-endpoint.sh` will do for local setups like kind.

If you use AKS, you should deploy the moose example into the cluster setting the appropriate labels (they show up in the log as well), as this script will most probably not work.

**Comment by bacongobbler** [comment]
Got it. Thanks for the context.

However, even with the script it still doesn't seem to be working. I'm not sure if the admission webhook is picking up on the new endpoint.

```
><> ./examples/assets/use-external-endpoint.sh
service/mooses-animals-com-admission-webhook patched
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
endpoints/mooses-animals-com-admission-webhook configured
><> k create -f examples/assets/manfred.yaml
Error from server (InternalError): error when creating "examples/assets/manfred.yaml": Internal error occurred: failed calling webhook "mooses.animals.com": Post "https://mooses-animals-com-admission-webhook.default.svc:443/?timeout=10s": dial tcp 10.96.78.189:443: connect: connection refused
```

Even if it did work, I don't think hard-coding to 172.17.0.1 will work for WSL2 as the network bridge from KinD to the host network is different for that operating system. We documented this process in the [HOWTO guide for Krustlet](https://github.com/deislabs/krustlet/blob/main/docs/howto/krustlet-on-kind.md#step-2-determine-the-default-gateway) if that helps.

For reference: the patch succeeded, and the endpoint exists.

```
><> k get svc mooses-animals-com-admission-webhook -o yaml | tail -n 12
spec:
  clusterIP: 10.96.78.189
  clusterIPs:
  - 10.96.78.189
  ports:
  - port: 443
    protocol: TCP
    targetPort: 8443
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
><> k get endpoints
NAME                                   ENDPOINTS         AGE
kubernetes                             172.18.0.2:6443   13m
mooses-animals-com-admission-webhook   172.17.0.1:8443   9m1s
```

**Comment by bacongobbler** [comment]
As a note: I don't think my comments here are blocking this from going through. I'm just providing my (relatively naive) feedback here to inform you how someone new to the feature is experiementing. Doc/tooling fixes shouldn't block this from going through IMO.

That being said, the concerns raised earlier from @kflansburg about requiring escalated privileges to register an admission webhook seems concerning. If one were to find a security flaw in an operator's admission webhook, the attacker now would have escalated cluster privileges. That seems like a very concerning attack vector.

Are there ways to avoid the need for privilege escalation? How do other operator frameworks register admission webhooks?

**Comment by kesselborn** [comment]
yeah ... testing this outside the cluster is a work around to speed up development and there will never be a really good solution for it. I just tried to reach parity with the [existing scripts in the current example code](https://github.com/deislabs/krustlet/blob/22486344c8a6a5924d8a8910f25bc1f171efc3bd/crates/krator/examples/assets/deploy.sh#L31-L39).

The correct way to test this is to deploy the moose operator into a Kubernetes cluster.

Regarding the security concerns: this is an example operator which is optimized for ease of use. You only need more privileges if you want to install necessary resources with the operator binary -- but that's not necessary. There are actually specific functions that will print out the manifests for you, so you can install the resources with `kubectl`, `helm`, etc.

Letting the operator produce these manifests is something I like as the operator and the documentation are not diverging -- you can always create the manifests that fit your operator.

**Review by kflansburg — CHANGES_REQUESTED** [review]
@kesselborn We have discussed and believe that it is not a good idea to include this security vulnerability as part of the demo, as it will encourage bad behavior. If you can change it to generate the YAML for the user to manually apply, and update the README, then I will approve. If you really want to keep this functionality, then I would ask that this specific behavior is gated behind a specific feature with `danger` in the name. 

@bacongobbler Were you able to get the Endpoint working? Was the issue with the gateway being used and WSL?

@thomastaylor312 You has also requested changes. Can you indicate whether those have been resolved?

**Comment by bacongobbler** [comment]
Thanks for the ping. In the end I was able to get it to work - just a quick misunderstanding of the networking setup on my end.

**Comment by kesselborn** [comment]
@kflansburg I do agree that this sets a bad precedent. I adjusted the example accordingly and added a comment to to the apply functions to note, that they should not be called in the operator.

**Comment by bacongobbler** [comment]
Providing some additional context as discussed offline...

We (Taylor and I) are maintainers on several large open source projects. Helm being one of them. In Helm 2, our stance was to provide a permissive default configuration. This allowed first-time users to start experimenting with Helm and Kubernetes without having to dive headfirst into the security controls. Unfortunately, this permissive configuration could grant a user a broad range of permissions they weren't intended to have. And for the most part users ignored the security documentation, instead choosing to criticize the project for "not doing the right thing" by default.

Learning from our mistakes, it's best to demonstrate "best practices" as the default rather than skewing towards ease-of-use. Especially when "ease-of-use" means skirting certain security controls.

Another great example of providing insecure defaults can be shown in [this article for MongoDB instances being hacked](https://www.zdnet.com/article/mongodb-ransacked-now-27000-databases-hit-in-mass-ransom-attacks/). The attackers hacked instances that did not have password-protected admin accounts. As in, they attacked MongoDB instances using the default settings.

As the Moose demonstration is the de-facto example for krator, I would not be surprised that many users will copy the example and apply it to their projects. Hence the concern.

File: crates/krator-derive/src/lib.rs
Comment (thomastaylor312) [reply crates/krator-derive/src/lib.rs:77]: I think we are ok for now. I'll do some double checking later

**Comment by kesselborn** [comment]
@bacongobbler thanks for the explanation -- I absolutely understand those points and do agree that this would have set a bad example :) 

Looking in the failed checks now

File: crates/krator/examples/moose.rs
Comment (kesselborn) [reply crates/krator/examples/moose.rs]: resolving this as this would mean pulling the operations, that need a lot of permissions, into krator

**Review by kflansburg — APPROVED** [review]
LGTM, thanks for your patience!

**Pull request merged**
